### PR TITLE
chore: update the readme file

### LINKED ISSUE #2775: Update the guidelines on how we test
## Motivation

The Readme file has outdated information on our testing infrastructure.

## Proposal

After merging #2748 , we should refresh the documentation.
Let's use a single section for Testing, either under Developing or standalone , and re-conciliate what's currently in Uni Tests and Linting and in Developing.

Be sure to distinguish between:
* Jest: unit tests for functions or small slices of code
* Storybook tests: rendering single components
* Cypress tests: integration testing with mocked API responses
* Acceptance tests: functional testing requiring the whole platform -- on the main Renku repository

### PR DESCRIPTION
**Pull request by lorenzo-cavazzi**
This PR restructures the Readme file to provide more consistent information to the developers.

* The initial `Development` section explains how to start developing code for the whole UI.
* The `UI client` and `UI server` have been restructured and expanded (especially the Client section). Grouping should be more intuitive going through the "tool stack", then the "code structure", then "testing" and finally mentioning other utilities useful for the development.

fix #2775

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,213 +95,217 @@ Kubectl and telepresence will allow you to inject code running on your developme
...
 
-The client is the [React-based](https://reactjs.org) front-end for RenkuLab. Development started in 2017, and, in the intervening years, we have striven to change our development style to reflect the evolving best practices around the tools we use. You will see that not all code conforms to the guidelines laid out in this document, but **new code** should follow these guidelines, and older code should be refactored at opportune moments, whenever possible, to conform as well.
+The UI client is the [React-based](https://reactjs.org) front-end for RenkuLab.
+Development started in 2017, and we have striven to change our development style
+to reflect the evolving best practices around the tools we use.
 
-Further technical details are available in [README file in the `client` folder](./client/README.md).
+Not all code conforms to the guidelines and best practices laid out in this
+document and you might find older code using deprecated technologies. that will
+be refactored in the future.
```
Comment (leafty) [inline README.md:106]: Typo?

```suggestion
document and you might find older code using deprecated technologies. That will
be refactored in the future.
```

File: README.md [hunk 2]
```diff
@@ -57,213 +95,217 @@ Kubectl and telepresence will allow you to inject code running on your developme
...
-The following example shows how to check for memory leaks:
+In short:
 
-    $ node --expose-gc ./node_modules/@craco/craco/dist/bin/jest --runInBand --logHeapUsage *
+- Global application state is kept in a single, global [Redux](https://redux.js.org) store.
+- Features should have
+  [slices](https://redux-toolkit.js.org/usage/usage-with-typescript#createslice) that
+  encapsulate the state they need and add the slices into the
+  [global store](https://redux-toolkit.js.org/api/configureStore).
+- Use the `useSelector` hook to access information from slices in components.
```
Comment (leafty) [inline README.md:143]:
```suggestion
- Use slice-specific selector hooks (e.g. [useWorkflowsSelector](https://github.com/SwissDataScienceCenter/renku-ui/blob/master/client/src/features/workflows/WorkflowsSlice.ts)), or the `useSelector` hook to access global state.
```

File: README.md [hunk 3]
```diff
@@ -57,213 +95,217 @@ Kubectl and telepresence will allow you to inject code running on your developme
...
-    $ npm install
-
-This will install the toolchain for developing the client as well as any libraries used in the UI.
+Based on [these suggestions from Redux](https://redux.js.org/faq/code-structure), we decided to use a few folders to bundle functions and components together, broadly following the "Feature folders" style.
 
-To run and interact with the UI code on your development machine, you can either use the Cypress tests, which provide mocked responses from the backend, or use telepresence to replace the UI client component in a K8s-based deployment with the component running on your machine.
+Here are the folders in `/client/src` where to place new components:
 
-### Cypress
+- `features`: create/use sub-folders to contain files identifying single features. These sometimes correspond to Renku abstractions, like "Projects", "Datasets", "Sessions", or to cross-entity features such as "Search" and "Dashboard". All RTK queries should be triggered by components in this folder. Most of the actions to save data in the Redux store slices should be defined here as well.
```
Comment (leafty) [inline README.md:156]: > All RTK queries should be triggered by components in this folder.

What does this mean? This sounds like a over restrictive rule.

File: README.md [hunk 4]
```diff
@@ -57,213 +95,217 @@ Kubectl and telepresence will allow you to inject code running on your developme
...
+- `components`: add here components that can be reused in different contexts. If something is clearly a shared component (E.G. `RenkuAlert`), put it here. If it's not obvious, and currently used by just one component, you can leave it in the `feature` folder (follow the principle: do not over-engineer it too early). Mind that we also store most of the temporary values in the Redux store, so you can define actions here if necessary.
+- `utils`: put here anything generic that doesn't fall into the previous categories (E.G. constants, helper functions, wrappers).
 
-We have an ever-growing suite of UI tests developed with Cypress. The tests can be run with mock responses from the backend, and can therefore be used to try out changes to the UI without access to a development cluster.
+Picking the perfect place isn't always straightforward and our current folder structure still has many outdated components that don't follow the convention. We plan to move them when already touching the code for other changes.
 
-    $ cd tests
-    $ npm install
-    $ npm run e2e:local
+#### **Use CSS modules**
```
Comment (ciyer) [inline README.md:162]: I think this should emphasize that the css modules are for styles that are local to a component.

```suggestion
#### **Use CSS modules for local styles**
```

File: README.md [hunk 5]
```diff
@@ -57,213 +95,217 @@ Kubectl and telepresence will allow you to inject code running on your developme
...
 
-#### Run Storybook
+We split testing into multiple categories. All these tests are required to pass
+before we merge a PR.
 
-Run the following command to start Storybook:
+#### Unit tests
 
-    $ npm run storybook
+We use [jest](https://jestjs.io) for unit testing.
```
Comment (ciyer) [inline README.md:175]:
```suggestion
We use [jest](https://jestjs.io) for **unit testing**.
```

File: README.md [hunk 6]
```diff
@@ -57,213 +95,217 @@ Kubectl and telepresence will allow you to inject code running on your developme
...
 
-This should automatically open your browser at http://localhost:6006 where you can see your components in action.
+Unit tests are used to test individual functions. We don't have a fixed rule for
+the amount of coverage we require, but we try to cover all the helper functions and
+the most important functions used by components.
```
Comment (ciyer) [inline README.md:179]:
```suggestion
the most important functions used by components. Note, we do not test components using jest/jsdom. Tests that require a browser/DOM environment are implemented either in Storybook or Cypress (see below).
```

File: README.md [hunk 7]
```diff
@@ -57,213 +95,217 @@ Kubectl and telepresence will allow you to inject code running on your developme
...
+Storybook files are named `*.stories.tsx` and refer to specific components.
+The overhead of writing stories is rather low, so we encourage you to write them
+for all new components.
+Whether to write stories for specific components depends on the size and re-usability
+of the component. The more simple and reusable a component is, the more important
+it is to have a story for it.
+Be sure to check out the full documentation [here](https://storybook.js.org/docs/react/writing-stories/introduction).
 
-Storybook's test runner transforms stories into executable tests, supporting Chromium and Firefox browsers. It ensures error-free rendering for non-play stories. For interactive stories with play functions, it validates both play function errors and successful assertions.
+You can run component tests using the following command:
```
Comment (ciyer) [inline README.md:206]:
```suggestion
You can run component tests using the following command in the client subfolder:
```

File: README.md [hunk 8]
```diff
@@ -57,213 +95,217 @@ Kubectl and telepresence will allow you to inject code running on your developme
...
+Cypress is also a great tool to run components in the browser and debug them. Most of the
+time, you don't need the whole platform to develop or update a component, so you can
+use Cypress to develop the UI without a dedicated RenkuLab deployment.
 
-The coding guidelines are detailed in full in [CODING_GUIDELINES.md](CODING_GUIDELINES.md).
+You can run Cypress tests using the following command:
 
-As explained above, not all the code in the repo conforms to these guidelines, but this section explains how code _should_ be developed in the repo.
+    $ cd tests
+    $ npm run e2e:local
```
Comment (ciyer) [inline README.md:262]: `:local` is not necessary
```suggestion
    $ npm run e2e
```

File: README.md [hunk 9]
```diff
@@ -57,213 +95,217 @@ Kubectl and telepresence will allow you to inject code running on your developme
...
-### **Use of CSS modules**
+    $ cd client
+    $ ./run-telepresence.sh
 
-[CSS modules](https://github.com/css-modules/css-modules) can be used to apply CSS styles locally and avoid leaking
-styles to the whole web application.
-[Create React App supports CSS modules out of the box](https://create-react-app.dev/docs/adding-a-css-modules-stylesheet).
+Telepresence replaces the UI client pod in the target Kubernetes instance. All the traffic
+is then redirected to a local process, making changes to files almost immediately available
+in your development RenkuLab instance.
```
Comment (ciyer) [inline README.md:306]:
```suggestion
in your development RenkuLab instance.

**Configuration**
The script allows configuration of certain aspects of the deployment through environment variables. Take a look at the script to see all the options that are available.

The script generates a `config.json` file into the `client/public` folder, and you can modify this file with a text editor and reload the browser to test out different configuration settings.
```

File: README.md [hunk 10]
```diff
@@ -325,94 +367,80 @@ flowchart LR
...
-    $ npm test
-    $ npm run lint
-
-Some linting errors can be automatically fixed by running
-`npm run lint-fix`.
+### Code structure
 
-We suggest using an IDE that supports eslint (like [VS Code](https://code.visualstudio.com) or similar) and configuring your IDE to integrate with our eslint configuration so that any linting errors will be displayed as you develop rather than waiting for the CI pipeline to flag them.
+We have an abstraction for storage in the `storage` folder. Currently, we support
+only Redis and we don't plan to use the UI server to store non-volatile data.
```
Comment (ciyer) [inline README.md:400]:
```suggestion
only Redis, and we do not currently use the UI server to store non-volatile data.
```

**Review by ciyer — CHANGES_REQUESTED** [review]
Nice improvement to the documentation! I have made some suggestions for clarification.

File: README.md (hunk 3 above)
Comment (lorenzo-cavazzi) [reply README.md]: That sounds indeed a bit misleading. I changed the text to emphasize that code interacting with the Redux store should go here in `*Api.ts` and `*Slice.ts` files.
See https://github.com/SwissDataScienceCenter/renku-ui/pull/2808/commits/ec811b333cc1d4969711d9452b37a81012340638

File: README.md (hunk 9 above)
Comment (lorenzo-cavazzi) [reply README.md:320]: Great addition, thanks!

File: README.md [hunk 11]
```diff
@@ -140,7 +140,9 @@ In short:
   [slices](https://redux-toolkit.js.org/usage/usage-with-typescript#createslice) that
   encapsulate the state they need and add the slices into the
   [global store](https://redux-toolkit.js.org/api/configureStore).
-- Use the `useSelector` hook to access information from slices in components.
+- Use slice-specific selector hooks (E.G.
```
Comment (leafty) [inline README.md:143]:
```suggestion
- Use slice-specific selector hooks (e.g.
```

see https://en.wiktionary.org/wiki/e.g.

File: README.md [hunk 12]
```diff
@@ -153,13 +155,25 @@ Based on [these suggestions from Redux](https://redux.js.org/faq/code-structure)
...
-- `utils`: put here anything generic that doesn't fall into the previous categories (E.G. constants, helper functions, wrappers).
-
-Picking the perfect place isn't always straightforward and our current folder structure still has many outdated components that don't follow the convention. We plan to move them when already touching the code for other changes.
-
-#### **Use CSS modules**
+- `features`: create/use sub-folders to contain files identifying single features.
+  These sometimes correspond to Renku abstractions, like "Projects", "Datasets",
+  "Sessions", or to cross-entity features such as "Search" and "Dashboard".
+  Wherever relevant, please add `*Api.ts` files containg the RTK queries and `*Slice.ts`
+  files for slices.
```
Comment (leafty) [inline README.md:162]: Per the coding guidelines, these should end in `.api.ts` and `.slice.ts`

```suggestion
  Wherever relevant, please add `*.api.ts` files containg the RTK queries and `*.slice.ts`
  files for slices.
```

File: README.md (hunk 11 above)
Comment (lorenzo-cavazzi) [reply README.md]: Changed the other occurrences too for consistency

**Review by ciyer — APPROVED** [review]
👍

**Review by leafty — APPROVED** [review]

**Pull request merged**